src/runtime_src/core/pcie/windows/device_windows.cpp

```diff
@@ -911,6 +932,62 @@ struct data_retention
 
 };
 
+struct mailbox
+{
+    using result_type = boost::any;
+
+    static xcl_mailbox
+        init_mailbox_info(const xrt_core::device* dev)
+    {
+        xcl_mailbox info = { 0 };
+        userpf::get_mailbox_info(dev->get_user_handle(), &info);
+        return info;
+    }
+
+    static result_type
+        get_info(const xrt_core::device* device, key_type key)
+    {
+        static std::map<const xrt_core::device*, xcl_mailbox> info_map;
+        static std::mutex mutex;
+        std::lock_guard<std::mutex> lk(mutex);
+
+        auto it = info_map.find(device);
+        if (it == info_map.end()) {
+            auto ret = info_map.emplace(device, init_mailbox_info(device));
+            it = ret.first;
+        }
+
+        const xcl_mailbox& info = (*it).second;
+        int i;
+        std::vector<std::string> vec;
+        switch (key) {
+        case key_type::mailbox_metrics:
+            vec.push_back((boost::format("raw bytes received: %d\n") %
+                           info.mbx_recv_raw_bytes).str());
+            for (i = 0; i < MAILBOX_REQ_MAX; i++)
+                vec.push_back(boost::str(boost::format("req[%d] received: %d\n")
+                                         % i % info.mbx_recv_req[i]));
+            return std::vector<std::string>(vec);
+        default:
+            throw std::runtime_error("device_windows::mailbox() unexpected qr "
```

Comment:
This c++code is not kernel 'c' code and should be written to be more readable.  This line can be broken into a string assignment and then thrown.